### PR TITLE
Initial support for the HIAC Liquid Particle Counter

### PR DESCRIPTION
**Pull request by chrisfriedline-benchling**
This PR adds initial support for the HIAC Liquid Particle Counter.

### CODE REVIEW

File: .vscode/settings.json
Comment (bdworth) [inline .vscode/settings.json:1]: Could you please revert these changes? If we want to make changes to IDE config files, that's fine, but it should be in a separate PR.

File: src/allotropy/allotrope/models/shared/definitions/definitions.py [hunk 1]
```diff
@@ -30,6 +30,8 @@
 
 TClass = str
 
+TIntValue = int
```
Comment (bdworth) [inline src/allotropy/allotrope/models/shared/definitions/definitions.py:33]: Is this used? I couldn't find a reference.

File: src/allotropy/allotrope/schema_parser/generate_schemas.py [hunk 2]
```diff
@@ -82,9 +82,11 @@ def generate_schemas(root_dir: Path) -> int:
         print(f"Generating models for schema: {rel_schema_path}...")  # noqa: T201
         schema_name = rel_schema_path.stem
         schema_path = os.path.join(root_dir, SCHEMA_DIR_PATH, rel_schema_path)
-        output_file = re.sub(
-            "/|-", "_", f"{rel_schema_path.parent}_{schema_name}.py"
-        ).lower()
+        output_file = (
+            re.sub("/|-", "_", f"{rel_schema_path.parent}_{schema_name}.py")
+            .replace(".embed.", "_embed_")
```
Comment (bdworth) [inline src/allotropy/allotrope/schema_parser/generate_schemas.py:87]: I trust this was needed, but where/why?

**Review by bdworth — CHANGES_REQUESTED** [review]
Thanks for working on this! I haven't finished reviewing but have some initial comments.

File: src/allotropy/allotrope/models/shared/definitions/definitions.py (hunk 1 above)
Comment (chrisfriedline-benchling) [reply src/allotropy/allotrope/models/shared/definitions/definitions.py:33]: It was used when I rolled my own model, before generating one from the embed schema. It could be removed, but it's a valid type and might be used at some point in the future.

File: src/allotropy/allotrope/schema_parser/generate_schemas.py (hunk 2 above)
Comment (chrisfriedline-benchling) [reply src/allotropy/allotrope/schema_parser/generate_schemas.py:87]: The model generation fails at the linting method with `.embed.` in the name, derived from `.embed.` in the json schema name. Ruff complained about an invalid name.

File: .vscode/settings.json
Comment (chrisfriedline-benchling) [reply .vscode/settings.json:1]: Sure thing.

File: src/allotropy/allotrope/schema_parser/generate_schemas.py (hunk 2 above)
Comment (nathan-stender) [reply src/allotropy/allotrope/schema_parser/generate_schemas.py:87]: This isn't specified in the current contributing / readme (as we don't expect outside contributors to be adding schemas just yet), but we just don't include the `embed` in the name (it's the same schema, just without external links)

Comment (chrisfriedline-benchling) [reply src/allotropy/allotrope/schema_parser/generate_schemas.py:87]: @nathan-stender Confirming that I should rename `light-obscuration.embed.schema.json` to `light-obscuration.schema.json`

Comment (nathan-stender) [reply src/allotropy/allotrope/schema_parser/generate_schemas.py:87]: Yup!

File: src/allotropy/allotrope/schemas/light-obscuration/REC/2021/12/light-obscuration.embed.schema.json [hunk 3]
```diff
@@ -0,0 +1,1342 @@
...
+              "type": "string",
+              "$asm.type": "http://www.w3.org/2000/01/rdf-schema#Class",
+              "$asm.lookup-property": "http://www.w3.org/2004/02/skos/core#prefLabel"
+            }
+          }
+        }
+      ]
+    }
+  },
+  "$defs": {
```
[the remark this answers is not in the document]
Comment (nathan-stender) [reply src/allotropy/allotrope/schemas/light-obscuration/REC/2021/12/light-obscuration.embed.schema.json:493]: Assuming the defs match (they should) yes.

The idea here is that while we need "embeded" schemas because our model generator code can't handle following links to other references at this time, the point of the shared code in the normal schema is that we shouldn't have to duplicate all this code in every schema.

File: tests/parsers/beckman_hiac/beckman_hiac_parser_test.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+from allotropy.parsers.utils.timestamp_parser import TimestampParser
+from allotropy.to_allotrope import allotrope_from_file
+
+
+@pytest.fixture()
+def test_file() -> Path:
+    f = Path(__file__).parent / "testdata/hiac_example_1.xlsx"
+    return f.absolute()
+
+
```
Comment (nathan-stender) [inline tests/parsers/beckman_hiac/beckman_hiac_parser_test.py:16]: We should have a test in here that checks the contents of the example json file, see the     `validate_schema` and `validate_contents` functions in other parser tests.

**Review by nathan-stender — CHANGES_REQUESTED** [review]
Great work! A couple requests.

File: src/allotropy/parsers/beckman_hiac/hiac_parser.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    def _extract_data(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Extract the Average data frame from the raw data. Initial use cases have focused on
+        only extracting the Average data, not the individual runs. The ASM does support multiple
+        Distribution objects, but they don't have names, so it's not possible to pick these out
+        after the fact. As such, this extraction only includes the Average data.
+
+        :param df: the raw dataframe
+        :return: the average data frame
+        """
+        start, end = self.get_data_index_bounds(df)
```
[the remark this answers is not in the document]
Comment (chrisfriedline-benchling) [reply src/allotropy/parsers/beckman_hiac/hiac_parser.py:62]: Done.

File: tests/parsers/beckman_hiac/beckman_hiac_parser_test.py (hunk 4 above)
Comment (chrisfriedline-benchling) [reply tests/parsers/beckman_hiac/beckman_hiac_parser_test.py:19]: For `validate_contents`, given that we have `NaN` in our outputs, I think we need to use `ignore_nan_inequality=True` in `DeepDiff`. In our example. We get the following.

As is:

```
DeepDiff(
    expected_dict,
    allotrope_dict,
    exclude_regex_paths=exclude_regex,
    ignore_type_in_groups=[(float, np.float64)],
)
```

Results:
```
{'values_changed': {"root['measurement document']['distribution document'][0]['distribution'][0]['differential particle density']['value']": {'new_value': nan,
   'old_value': nan},
  "root['measurement document']['distribution document'][0]['distribution'][0]['differential count']['value']": {'new_value': nan,
   'old_value': nan},
  "root['measurement document']['distribution document'][0]['distribution'][1]['differential particle density']['value']": {'new_value': nan,
   'old_value': nan},
  "root['measurement document']['distribution document'][0]['distribution'][1]['differential count']['value']": {'new_value': nan,
   'old_value': nan},
  "root['measurement document']['distribution document'][0]['distribution'][2]['differential particle density']['value']": {'new_value': nan,
   'old_value': nan},
  "root['measurement document']['distribution document'][0]['distribution'][2]['differential count']['value']": {'new_value': nan,
   'old_value': nan},
  "root['measurement document']['distribution document'][0]['distribution'][3]['differential particle density']['value']": {'new_value': nan,
   'old_value': nan},
  "root['measurement document']['distribution document'][0]['distribution'][3]['differential count']['value']": {'new_value': nan,
   'old_value': nan},
  "root['measurement document']['distribution document'][0]['distribution'][4]['differential particle density']['value']": {'new_value': nan,
   'old_value': nan},
  "root['measurement document']['distribution document'][0]['distribution'][4]['differential count']['value']": {'new_value': nan,
   'old_value': nan}}}

```

However, if we add that param to `DeepDiff`, we can pass

```
DeepDiff(
    expected_dict,
    allotrope_dict,
    exclude_regex_paths=exclude_regex,
    ignore_type_in_groups=[(float, np.float64)],
)
```

Results:
```
{}
```

Thoughts @nathan-stender?

Comment (nathan-stender) [reply tests/parsers/beckman_hiac/beckman_hiac_parser_test.py:19]: `ignore_nan_inequality=True` seems like a fine adjustment to me.

 However, I'm kind of surprised `NaN` passes schema validation. I'm no expert, but it seems like most things I look at say that `NaN` is not valid JSON. It seems like this is a bug in the validator we use (https://github.com/tdegrunt/jsonschema/issues/53)

I would say we should just omit fields when their value is `NaN`, but it seems like at least one of these (`differential count`) is required in the schema. This is possibly a question for @LimulusPolyphemus

File: src/allotropy/allotrope/models/light_obscuration_rec_2021_12_light_obscuration.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+)
+from allotropy.allotrope.models.shared.definitions.definitions import (
+    TBooleanValue,
+    TDateTimeValue,
+    TStringValue,
+)
+
+
+@dataclass
+class RepetitionSetting:
```
Comment (nathan-stender) [inline src/allotropy/allotrope/models/light_obscuration_rec_2021_12_light_obscuration.py:24]: This value is `tIntValue`. Seems like this is something we should add to `definitions` (like `tStringValue`)

File: src/allotropy/allotrope/models/shared/definitions/definitions.py (hunk 1 above)
Comment (nathan-stender) [reply src/allotropy/allotrope/models/shared/definitions/definitions.py:33]: I think with the re-structuring to use references this would be used again for `repetition setting`. Just need to add it with ref to `tIntValue` and then add it to `definitions.json`

Comment (chrisfriedline-benchling) [reply src/allotropy/allotrope/models/shared/definitions/definitions.py]: Added those changes for repetition setting.

File: src/allotropy/allotrope/models/light_obscuration_rec_2021_12_light_obscuration.py (hunk 6 above)
Comment (chrisfriedline-benchling) [reply src/allotropy/allotrope/models/light_obscuration_rec_2021_12_light_obscuration.py]: Done.

File: tests/parsers/beckman_hiac/beckman_hiac_parser_test.py (hunk 4 above)
Comment (chrisfriedline-benchling) [reply tests/parsers/beckman_hiac/beckman_hiac_parser_test.py:19]: As discussed, I've updated the model to make these two missing fields optional. We will regroup with Allotrope as we work to enhance this model.

**Comment by chrisfriedline-benchling** [comment]
> @chrisfriedline-benchling is possible to confirm whether this output is coming from the PharmaSpec software?
> 
> PharmaSpec is the software that is used in conjunction with the HAIC particle counter. Keeping with the adapter naming convention this should be the "PharmaSpec" adapter rather than the "HAIC particle counter" adapter.

@LimulusPolyphemus - it's PharmSpec v3.0.

**Review by LimulusPolyphemus — DISMISSED** [review]
Yes, that looks good to me.

File: src/allotropy/parsers/beckman_pharmspec/pharmspec_parser.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+def get_property_from_sample(property_name: str, value: Any) -> Any:
+    return property_lookup[property_name](value=value)
+
+
+class PharmSpecParser(VendorParser):
+    def to_allotrope(self, named_file_contents: NamedFileContents) -> Model:
+        df = pd.read_excel(named_file_contents.contents, header=None, engine="openpyxl")
+        return self._get_model(df)
+
+    def _get_model(self, df: pd.DataFrame) -> Model:
```
Comment (nathan-stender) [inline src/allotropy/parsers/beckman_pharmspec/pharmspec_parser.py:48]: Nit: doesn't seem like you need the redirect to `_setup_model` (perhaps just rename `_setup_model` to `_get_model` and remove this?)

**Review by nathan-stender — DISMISSED** [review]

File: src/allotropy/parsers/beckman_pharmspec/pharmspec_parser.py (hunk 7 above)
Comment (chrisfriedline-benchling) [reply src/allotropy/parsers/beckman_pharmspec/pharmspec_parser.py]: Done.

File: src/allotropy/parsers/beckman_pharmspec/pharmspec_parser.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    ) -> list[DistributionDocumentItem]:
+        """Create the distribution document. First, we create the actual distrituion, which itself
+        contains a list of DistributionDocumentItem objects. The DistributionDocumentItem objects represent the values
+        from the rows of the incoming dataframe.
+
+        If we were able to support more than one data frame instead of just the average data, we could
+        return a DistributionDocument with more than one item. For the use cases we've seen, there is
+        only a single Distribution being returned at this time, containing the average data.
+
+        :param df: The average datafreame
```
Comment (nathan-stender) [inline src/allotropy/parsers/beckman_pharmspec/pharmspec_parser.py:114]:
```suggestion
        :param df: The average dataframe
```

File: tests/parsers/beckman_pharmspec/testdata/hiac_example_1.json [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+                {
+                    "detector identifier": "",
+                    "detector model number": "HRLD150@10ml"
+                }
+            ]
+        },
+        "calculated data aggregate document": {
+            "calculated data document": [
+                {
+                    "calculated data name": "Average",
```
Comment (nathan-stender) [inline tests/parsers/beckman_pharmspec/testdata/hiac_example_1.json:264]: Typically, calculated data documents have a "calculated data identifier" (a UUID) and a data source document that link them back to the measurements they are derived from.

@LimulusPolyphemus thoughts?

Comment (LimulusPolyphemus) [reply tests/parsers/beckman_pharmspec/testdata/hiac_example_1.json:264]: @chrisfriedline-benchling - @nathan-stender is correct, the "calculated data identifier" should always be included so the "calculated result" can be referenced elsewhere if needed.

**Comment by LimulusPolyphemus** [comment]
@chrisfriedline-benchling going back to the old comment, we should update the name to beckman_pharmspec for consistency with the other adapters.

File: tests/parsers/beckman_pharmspec/testdata/hiac_example_1.json (hunk 9 above)
Comment (LimulusPolyphemus) [reply tests/parsers/beckman_pharmspec/testdata/hiac_example_1.json:264]: @chrisfriedline-benchling looking at this again - the `calculate data documents` should also include the `data source aggregate document` with a `data source documents` with the `data source identifier` field to cite the `measurement identifier` from which the input data for the Average is derived.

**Review by LimulusPolyphemus — CHANGES_REQUESTED** [review]
@chrisfriedline-benchling let me know if it is helpful to meet and discuss the `calculated data document` and `data source document`

File: src/allotropy/parsers/beckman_pharmspec/pharmspec_parser.py (hunk 8 above)
Comment (chrisfriedline-benchling) [reply src/allotropy/parsers/beckman_pharmspec/pharmspec_parser.py]: fixed.

**Comment by chrisfriedline-benchling** [comment]
> @chrisfriedline-benchling going back to the old comment, we should update the name to beckman_pharmspec for consistency with the other adapters.

Done.

**Comment by chrisfriedline-benchling** [comment]
> @chrisfriedline-benchling let me know if it is helpful to meet and discuss the `calculated data document` and `data source document`

I think this should be good now.

**Comment by chrisfriedline-benchling** [comment]
> @chrisfriedline-benchling the `calculated data documents` look good.
> 
> The only issues I see are that (i) The `data source feature` should be the keyword for the value being referenced as an input (e.g. particle size, cumulative count, cumulative particle density)
> 
> (ii) The `calculated data name` should include the label column header for the average, e.g. "Average_Particle_Size" instead of just "Average" since this will present challenges for distinguishing between the Average values when converting these back to a .csv

Fixed.

**Review by LimulusPolyphemus — APPROVED** [review]

File: src/allotropy/parsers/beckman_pharmspec/beckman_pharmspec_parser.py [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+                    MeasurementDocumentItem(
+                        measurement_identifier=name,
+                        measurement_time=pd.to_datetime(
+                            str(df.at[8, 5]).replace(".", "-")
+                        ).isoformat(timespec="microseconds")
+                        + "Z",
+                        device_control_aggregate_document=DeviceControlAggregateDocument(
+                            device_control_document=[
+                                DeviceControlDocumentItemModel(
+                                    flush_volume_setting=TQuantityValueMilliliter(0),
```
Comment (nathan-stender) [inline src/allotropy/parsers/beckman_pharmspec/beckman_pharmspec_parser.py:197]: Just confirming: we are purposely setting this to 0 because it is required by the schema, but not provided?

**Review by nathan-stender — APPROVED** [review]

**Review by nathan-stender — COMMENTED** [review]
Can you please follow up with a CHANGE_NOTES on light obscuration schema that notes the base ASM schema and the diffs? Or perhaps an ask for @LimulusPolyphemus ?

File: src/allotropy/parsers/beckman_pharmspec/beckman_pharmspec_parser.py (hunk 10 above)
Comment (chrisfriedline-benchling) [reply src/allotropy/parsers/beckman_pharmspec/beckman_pharmspec_parser.py:197]: That's correct.
